# æå–åŸºå› ç»„ä¸­çš„å†…å«å­ã€å¤–æ˜¾å­ä»¥åŠåŸºå› é—´åŒº
æ–‡ç« æ¥è‡ªï¼šhttps://www.jieandze1314.com/post/cnposts/165/ ï¼šä»…ä¾›æœ¬äººå­¦ä¹ å‚è€ƒ
## å‰è¨€

<div align=center>
<img src="https://user-images.githubusercontent.com/71922803/187844931-0e1ee826-be81-4f85-aa33-ad1c82502c4c.png" width="450">
</div>


DaveTangçš„è¿™ç¯‡åšå®¢æ›´æ–°äº2014å¹´ï¼Œé‚£æ—¶è½¬å½•ç»„æµ‹åºå¾ˆç«çƒ­ã€‚RNA-Seqå¾€å¾€æ˜¯æŠŠreadsæ¯”å¯¹å›åŸºå› ç»„æˆ–è½¬å½•ç»„ï¼Œç„¶åå°±æ˜¯å¯¹readsè¿›è¡Œæ³¨é‡Šï¼Œçœ‹çœ‹å®ƒä»¬è½åœ¨ä»€ä¹ˆä½ç½®ï¼Œå…¶å®å®šé‡ä¹Ÿæ˜¯å…¶ä¸­ä¸€æ­¥ã€‚è¿™é‡Œï¼Œä½œè€…å°±å¸¦æˆ‘ä»¬çœ‹çœ‹ä»–æ˜¯å¦‚ä½•è·å¾—å¤–æ˜¾å­ï¼ˆexonicï¼‰ã€å†…å«å­ï¼ˆintronicï¼‰ã€åŸºå› é—´åŒºï¼ˆintergenic regionsï¼‰çš„åæ ‡çš„ã€‚
## é¦–å…ˆä¸‹è½½å‚è€ƒåŸºå› ç»„æ³¨é‡Š
```
#hg38ç‰ˆæœ¬
wget -c ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_41/gencode.v41.annotation..gz

zcat gencode.v41.annotation.gtf.gz | head -n6 
##description: evidence-based annotation of the human genome (GRCh38), version 41 (Ensembl 107)
##provider: GENCODE
##contact: gencode-help@ebi.ac.uk
##format: gtf
##date: 2022-05-12
chr1    HAVANA  gene    11869   14409   .       +       .       gene_id "ENSG00000223972.5"; gene_type "transcribed_unprocessed_pseudogene"; gene_name "DDX11L1"; level 2; hgnc_id "HGNC:37102"; havana_gene "OTTHUMG00000000961.2";
```
å…¶ä¸­çš„ç¬¬ä¸‰åˆ—è¯´æ˜äº†featureçš„ç±»å‹ï¼Œå…¶ä¸­ä¼šæœ‰exonã€CDSã€UTRç­‰ç­‰
```
zcat gencode.v44.annotation.gtf.gz| grep -v "^#" | cut -f3 | sort | uniq -c | sort -k1rn
1625321 exon
 871490 CDS
 377983 UTR
 251236 transcript
  97009 start_codon
  90749 stop_codon
  61852 gene
    119 Selenocysteine
```
å®ƒä»¬ä¹‹é—´çš„å…³ç³»ï¼Œå¯ä»¥çœ‹çœ‹ï¼šhttp://blog.sciencenet.cn/blog-1271266-792469.html ä»¥åŠ http://www.dxy.cn/bbs/topic/36728037
UTRï¼ˆUntranslatedRegions)å³éç¿»è¯‘åŒºï¼Œæ˜¯ä¿¡ä½¿RNAï¼ˆmRNAï¼‰åˆ†å­ä¸¤ç«¯çš„éç¼–ç ç‰‡æ®µï¼Œ`UTRåœ¨DNAåºåˆ—ä¸­ç®—æ˜¯å¤–æ˜¾å­`;5â€™-UTRä»mRNAèµ·ç‚¹çš„ç”²åŸºåŒ–é¸Ÿå˜Œå‘¤æ ¸è‹·é…¸å¸½å»¶ä¼¸è‡³AUGèµ·å§‹å¯†ç å­ï¼Œ3â€™-UTRä»ç¼–ç åŒºæœ«ç«¯çš„ç»ˆæ­¢å¯†ç å­å»¶ä¼¸è‡³å¤šèšAå°¾å·´ï¼ˆPoly-Aï¼‰çš„æœ«ç«¯ï¼›
åŸºå› é—´åŒºæ˜¯æ¯ä¸ªåŸºå› ä¹‹é—´çš„é—´éš”åºåˆ—ï¼Œä¸å±äºå¤–æ˜¾å­ä¹Ÿä¸å±äºå†…å«å­ï¼Œå®ƒæ˜¯non codingï¼›

<div align=center>
<img src="https://user-images.githubusercontent.com/71922803/187847850-3b5d9a82-d988-49af-b303-775c5456a26d.png" width="450">
</div>

>ã€å›¾ç‰‡ä¸­æè¿°çš„è¿‡ç¨‹ï¼šåŸºå› ç»è¿‡è½¬å½•å½¢æˆ pre-mRNAï¼Œè¿™é‡Œé¢åŒ…å«ç€å†…å«å­å’Œå¤–æ˜¾å­ï¼ˆ5ç«¯æ˜¯ä¸€ä¸ªå¤–æ˜¾å­ï¼Œä½†æ˜¯è¿™æ®µå¤–æ˜¾å­ä¸ä»…åŒ…å«CDSï¼Œè¿˜åŒ…å«5â€™ UTRï¼›3ç«¯ä¹Ÿæ˜¯ä»¥å¤–æ˜¾å­ç»“æŸï¼Œä½†å®ƒä¹Ÿæ˜¯ä¸ä»…åŒ…å«CDSï¼Œè¿˜åŒ…å«3â€™ UTRï¼‰ï¼Œç»è¿‡å‰ªæ¥å½¢æˆæˆç†ŸmRNA,å†…å«å­å·²å‡æ‰ã€‚å¦‚æœæŠ›å¼€åæ¥åŠ ä¸Šå»çš„capå’Œpoly Açš„è¯ï¼Œè¿™æ—¶å…¨æ˜¯å¤–æ˜¾å­ï¼Œä½†æ˜¯ä¸å…¨æ˜¯CDSï¼Œå› ä¸ºåªæœ‰ä¸­é—´çš„é‚£éƒ¨åˆ†ä»¥èµ·å§‹å¯†ç å­å¼€å§‹ã€ä»¥ç»ˆæ­¢å¯†ç å­ç»“æŸçš„ç‰‡æ®µæ‰æ˜¯CDSï¼Œåªæœ‰è¿™éƒ¨åˆ†æ‰ä¼šè¢«ç¿»è¯‘æˆè›‹ç™½è´¨ã€‘

> **è½¬å½•æœ¬ï¼ˆtranscript)**æ˜¯åŸºå› é€šè¿‡è½¬å½•å½¢æˆçš„ä¸€ç§æˆ–å¤šç§å¯ä¾›ç¼–ç è›‹ç™½è´¨çš„æˆç†Ÿçš„mRNAã€‚ä¸€ä¸ªåŸºå› ç”±äºå¯å˜å‰ªåˆ‡ç°è±¡ï¼Œå¯èƒ½äº§ç”Ÿå¤šä¸ªè½¬å½•æœ¬ã€‚ã€åŸºå› è½¬å½•çš„è¿‡ç¨‹ï¼Œé¦–å…ˆæ˜¯å½¢æˆå‰ä½“mRNAï¼Œé€šè¿‡å‰ªåˆ‡å†…å«å­è¿æ¥å¤–æ˜¾å­ï¼Œ5â€™ç«¯åŠ å¸½åŠ3â€™ç«¯åŠ å°¾ä¹‹åå½¢æˆæˆç†Ÿçš„mRNAã€‚å¯å˜å‰ªåˆ‡å°±æ˜¯ï¼šåœ¨å‰ªåˆ‡çš„è¿‡ç¨‹ä¸­å¯èƒ½ä¼šå‰ªåˆ‡æ‰å¤–æ˜¾å­ï¼Œä¹Ÿæœ‰å¯èƒ½ä¿ç•™éƒ¨åˆ†å†…å«å­ï¼Œå› æ­¤ä¼šå½¢æˆå¤šç§mRNAå³å¤šä¸ªè½¬å½•æœ¬ã€‘ï¼Œè€Œæˆ‘ä»¬å¹³æ—¶ç ”ç©¶æŸä¸ªåŸºå› çš„åŠŸèƒ½ï¼Œå®é™…æ˜¯ç ”ç©¶å®ƒçš„ä¸€ä¸ªè½¬å½•æœ¬ç¼–ç çš„è›‹ç™½çš„åŠŸèƒ½ã€‚ä¸€èˆ¬æƒ…å†µä¸‹å®ƒçš„ä¸åŒçš„è½¬å½•æœ¬åˆ†å¸ƒåœ¨ä¸åŒç±»å‹çš„ç»†èƒä¸­ï¼Œå½“ç„¶ä¹Ÿæœ‰å¯èƒ½å¤šç§è½¬å½•æœ¬åŒæ—¶å­˜åœ¨äºæŸä¸€ç»†èƒä¸­ã€‚
>ç¼–ç åŒºï¼ˆCoding regionï¼ŒCDSï¼‰ æ˜¯mRNAåºåˆ—ä¸­ç¼–ç è›‹ç™½è´¨çš„é‚£éƒ¨åˆ†åºåˆ—
## è·å¾—å¤–æ˜¾å­ã€å†…å«å­ä»¥åŠåŸºå› é—´åŒº
### å¤–æ˜¾å­çš„è·å¾— => mergeBed
ä»GTFæ–‡ä»¶ä¸­å…¶å®å·²ç»èƒ½æ‰¾åˆ°exonçš„è¸ªè¿¹ï¼Œä½†æ˜¯åŒä¸€ä¸ªåŸºå› ä¸‹çš„å„ä¸ªexonå¯èƒ½ä¼šæœ‰äº¤é›†ï¼Œäºæ˜¯è¿™é‡Œä½œè€…åœ¨ç»Ÿè®¡æ—¶ï¼Œå°†å­˜åœ¨é‡å çš„å¤–æ˜¾å­èåˆåœ¨ä¸€èµ·ï¼ˆè¿™é‡Œåªæ˜¯é‡‡ç”¨æœ€ç®€å•çš„æ–¹æ³•è¿›è¡Œæ¢ç´¢ï¼Œä½†ä¸æ˜¯æœ€å‡†ç¡®çš„ï¼‰
<div align=center>
<img src="https://user-images.githubusercontent.com/71922803/187849154-f6bde10d-d0da-403e-a1af-4d6d5dcdf0c8.png" width="450">
</div>

ä½¿ç”¨`bedtools`çš„`mergeBed`åŠŸèƒ½ï¼Œä½†å‰ææ˜¯å…ˆä½¿ç”¨`sortBed`å°†åæ ‡è¿›è¡Œæ’åº

```
# é¦–å…ˆæŒ‘å‡ºexonçš„èµ·å§‹å’Œç»ˆæ­¢ä½ç‚¹
zcat gencode.v33.annotation.gtf.gz | awk 'BEGIN{OFS="\t";} $3=="exon" {print $1,$4-1,$5}' | head -n5
chr1	11868	12227
chr1	12612	12721
chr1	13220	14409
chr1	12009	12057
chr1	12178	12227
# ç„¶åè¿›è¡Œæ’åºï¼Œæ¥ç€è¿›è¡Œmerge
zcat gencode.v33.annotation.gtf.gz | awk 'BEGIN{OFS="\t";} $3=="exon" {print $1,$4-1,$5}' | sortBed | mergeBed -i - | gzip >gencode.v33.exon.annotation.gtf.gz

zcat gencode.v33.exon.annotation.bed.gz | head -n5
chr1	11868	12227
chr1	12612	12721
chr1	12974	13052
chr1	13220	14501
chr1	15004	15038
```
>å°tipï¼š ä¸Šé¢ğŸ‘†çš„awkå‘½ä»¤ä¸­å‡ºç°äº†ä¸€ä¸ª{print $1,$4-1,$5}ï¼Œå…¶ä¸­$4æ˜¯æŒ‡çš„GTFçš„ç¬¬4åˆ—ï¼Œå³èµ·å§‹ä½ç‚¹ã€‚å®ƒè¦å‡ä¸€æ˜¯å› ä¸ºè¦ä»GTFå˜ä¸ºBEDï¼Œè€Œå®ƒä»¬çš„åæ ‡ç³»ç»Ÿä¸ä¸€è‡´ï¼šGTF has 1-based start coordinates, and BED has 0-based start coordinates

### å†…å«å­çš„è·å¾— => subtractBed
<div align=center>
<img src="https://user-images.githubusercontent.com/71922803/187851318-9ffa4ff4-8fd2-4657-b0d1-84814248f42b.png" width="450">
</div>

```
# é€‰æ‹©GTFä¸­geneçš„featureï¼Œç„¶åæ’é™¤merged exon
zcat gencode.v33.annotation.gtf.gz | awk 'BEGIN{OFS="\t";} $3=="gene" {print $1,$4-1,$5}' | sortBed | subtractBed -a stdin -b gencode.v33.exon.annotation.bed.gz | gzip >gencode.v33.intron.annotation.bed.gz

zcat gencode.v33.intron.annotation.bed.gz | head -n5
# ç»“æœå¯ä»¥å’Œä¸Šé¢ğŸ‘†çš„exonç»“æœå¯¹ç…§
chr1	12227	12612 
chr1	12721	12974
chr1	13052	13220
chr1	14501	15004
chr1	15038	15795
```
### åŸºå› é—´åŒºçš„è·å¾— => complementBed

<div align=center>
<img src="https://user-images.githubusercontent.com/71922803/187851565-67827656-4875-4f35-bdc7-bb04ecdd724c.png" width="450">
</div>

éœ€è¦å…ˆè·å¾—æ•´ä¸ªåŸºå› ç»„å„ä¸ªæŸ“è‰²ä½“çš„é•¿åº¦ï¼Œç„¶åä»ä¸­æ’é™¤æ‰gene

```
# å„ä¸ªæŸ“è‰²ä½“çš„é•¿åº¦(åœ¨bedtools complementBedçš„å¸®åŠ©æ–‡æ¡£ä¸­)
mysql --user=genome --host=genome-mysql.cse.ucsc.edu -A -e         "select chrom, size from hg38.chromInfo" >hg38.genome
# å»æ‰hg38.genomeç¬¬ä¸€è¡Œæ³¨é‡Šï¼šchrom	size
 sed -i '1d' hg38.genome
# æ’é™¤gene
zcat gencode.v33.annotation.gtf.gz | awk 'BEGIN{OFS="\t";} $3=="gene" {print $1,$4-1,$5}' | sortBed | complementBed -i stdin -g hg38.genome | gzip >gencode.v33.intergenic.annotation.bed.gz
```
**ä½†è¿™é‡ŒæŠ¥äº†ä¸ªé”™ï¼šError: Sorted input specified, but the file stdin has the following record with a different sort order than the genomeFile hg38.genome**
**å®é™…ä¸Šé—®é¢˜å°±æ˜¯å› ä¸ºæˆ‘ä»¬çš„sort bedåçš„chrxxä¸æ˜¯æŒ‰ç…§æ•°å­—æ’åºçš„ï¼ˆhttp://asearchforsolutions.blogspot.com/2018/11/error-sorted-input-specified-but-file.html**
```
zcat gencode.v33.annotation.gtf.gz | awk 'BEGIN{OFS="\t";} $3=="gene" {print $1,$4-1,$5}' | sortBed | cut -f1 | uniq
chr1
chr10
chr11
chr12
chr13
chr14
chr15
chr16
chr17
chr18
# è€Œè¿™é‡Œçš„hg38.genomeå´æ˜¯æ‹å¥½é¡ºåºçš„
cat hg38.genome | cut -f1 | uniq | head
chr1
chr2
chr3
chr4
chr5
chr6
chr7
chrX
chr8
chr9
# å› æ­¤æˆ‘ä»¬â€œæƒå¼ºå‡Œå¼±â€ï¼Œå°†hg38.genomeæ”¹æˆå’ŒsortBedåä¸€è‡´çš„ï¼ˆç›¸æ¯”ä¹‹ä¸‹hg38.genomeæ›´å®¹æ˜“ä¿®æ”¹ï¼‰ï¼Œä¹Ÿä¸æŒ‰ç…§å­—æ¯æ’åº
cat hg38.genome| sort -k1.4 >hg38-2.genome
head hg38-2.genome
chr1
chr10
chr10_GL383545v1_alt
chr10_GL383546v1_alt
chr10_KI270824v1_alt
chr10_KI270825v1_alt
chr10_KN196480v1_fix
chr10_KN538365v1_fix
chr10_KN538366v1_fix
chr10_KN538367v1_fix
```
äºæ˜¯ï¼Œç»§ç»­è¿›è¡Œä»£ç ï¼Œå°±æ²¡æœ‰é—®é¢˜å•¦ï¼š
```
zcat gencode.v33.annotation.gtf.gz | awk 'BEGIN{OFS="\t";} $3=="gene" {print $1,$4-1,$5}' | sortBed | complementBed -i stdin -g hg38-2.genome | gzip >gencode.v33.intergenic.annotation.bed.gz

zcat gencode.v33.intergenic.annotation.bed.gz | head -n5
chr1	0	11868
chr1	31109	34553
chr1	36081	52472
chr1	53312	57597
chr1	64116	65418
```
**åˆ°æ­¤ä¾¿æˆåŠŸæå–åŸºå› ç»„ä¸­çš„å†…å«å­ã€å¤–æ˜¾å­ä»¥åŠåŸºå› åŒºé—´**
## è·å¾—å¯¹åº”çš„.faæ–‡ä»¶
1.ä¸‹è½½äººç±»åŸºå› ç»„åºåˆ—
2.ä½¿ç”¨bedtools
```
Tool:    bedtools getfasta (aka fastaFromBed)
Version: v2.30.0
Summary: Extract DNA sequences from a fasta file based on feature coordinates.

Usage:   bedtools getfasta [OPTIONS] -fi <fasta> -bed <bed/gff/vcf>

Options:
        -fi             Input FASTA file
        -fo             Output file (opt., default is STDOUT
        -bed            BED/GFF/VCF file of ranges to extract from -fi
        -name           Use the name field and coordinates for the FASTA header
        -name+          (deprecated) Use the name field and coordinates for the FASTA header
        -nameOnly       Use the name field for the FASTA header
        -split          Given BED12 fmt., extract and concatenate the sequences
                        from the BED "blocks" (e.g., exons)
        -tab            Write output in TAB delimited format.
        -bedOut         Report extract sequences in a tab-delimited BED format instead of in FASTA format.
                        - Default is FASTA format.
        -s              Force strandedness. If the feature occupies the antisense,
                        strand, the sequence will be reverse complemented.
                        - By default, strand information is ignored.
        -fullHeader     Use full fasta header.
                        - By default, only the word before the first space or tab
                        is used.
        -rna    The FASTA is RNA not DNA. Reverse complementation handled accordingly.
```
## åœ¨æå–åŸºå› é—´åŒºåºåˆ—è¿‡ç¨‹ä¸­å‡ºç°å¦‚ä¸‹é—®é¢˜

```
WARNING. chromosome (chr10_GL383545v1_alt) was not found in the FASTA file. Skipping.
WARNING. chromosome (chr10_GL383546v1_alt) was not found in the FASTA file. Skipping.
WARNING. chromosome (chr10_KI270824v1_alt) was not found in the FASTA file. Skipping.
WARNING. chromosome (chr10_KI270825v1_alt) was not found in the FASTA file. Skipping.
WARNING. chromosome (chr10_KN196480v1_fix) was not found in the FASTA file. Skipping.
WARNING. chromosome (chr10_KN538365v1_fix) was not found in the FASTA file. Skipping.
WARNING. chromosome (chr10_KN538366v1_fix) was not found in the FASTA file. Skipping.
```

![image](https://user-images.githubusercontent.com/71922803/187893910-7c99ba00-0068-4888-a856-9f3e309707b2.png)

ç½‘ä¸Šæ‰¾åˆ°å…¶ä»–æ•™ç¨‹è¯´æ˜¯åŸºå› ç¢ç‰‡ï¼Œå½±å“åº”è¯¥ä¸å¤§
å¯ä»¥æ ¹æ®ä¸Šè¿°å›¾ç‰‡å°†å…¶å‰”é™¤
